File: py/google/fhir/r4/references_test.py
#
# Copyright 2020 Google LLC
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#      http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
"""Tests for the references module."""

from absl.testing import absltest
from proto.google.fhir.proto.r4.core import datatypes_pb2
from google.fhir import references
from google.fhir.utils import proto_utils


class ReferencesTest(absltest.TestCase):
  """A suite of tests to ensure proper FHIR reference handling."""

  def testSplitIfRelativeReference_withRelativeReference_succeeds(self):
    ref = datatypes_pb2.Reference(
        uri=datatypes_pb2.String(value='Practitioner/example'),
        display=datatypes_pb2.String(value='Dr Adam Careful'))

    uri_field = ref.DESCRIPTOR.fields_by_name['uri']
    practitioner_id_field = ref.DESCRIPTOR.fields_by_name['practitioner_id']
    self.assertTrue(proto_utils.field_is_set(ref, uri_field))
    self.assertFalse(proto_utils.field_is_set(ref, practitioner_id_field))

    references.split_if_relative_reference(ref)

    self.assertFalse(proto_utils.field_is_set(ref, uri_field))
    self.assertTrue(proto_utils.field_is_set(ref, practitioner_id_field))
    self.assertEqual(
        proto_utils.get_value_at_field(ref, practitioner_id_field),
        datatypes_pb2.ReferenceId(value='example'))

  def testSplitIfRelativeReference_withFragmentReference_succeeds(self):
    ref = datatypes_pb2.Reference(uri=datatypes_pb2.String(value='#org-1'))

    uri_field = ref.DESCRIPTOR.fields_by_name['uri']
    fragment_field = ref.DESCRIPTOR.fields_by_name['fragment']
    self.assertTrue(proto_utils.field_is_set(ref, uri_field))
    self.assertFalse(proto_utils.field_is_set(ref, fragment_field))

    references.split_if_relative_reference(ref)

    self.assertFalse(proto_utils.field_is_set(ref, uri_field))
    self.assertTrue(proto_utils.field_is_set(ref, fragment_field))
    self.assertEqual(
        proto_utils.get_value_at_field(ref, fragment_field),
        datatypes_pb2.String(value='org-1'))

  def testSplitIfRelativeReference_withUrlScheme_succeeds(self):
    ref = datatypes_pb2.Reference(
        uri=datatypes_pb2.String(
            value='http://acme.com/ehr/fhir/Practitioner/2323-33-4'))

    references.split_if_relative_reference(ref)

    uri_field = ref.DESCRIPTOR.fields_by_name['uri']
    self.assertEqual(
        proto_utils.get_value_at_field(ref, uri_field),
        datatypes_pb2.String(
            value='http://acme.com/ehr/fhir/Practitioner/2323-33-4'))

if __name__ == '__main__':
  absltest.main()

File: cc/google/fhir/references_test.cc
// Copyright 2020 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#include "google/fhir/references.h"

#include "gmock/gmock.h"
#include "gtest/gtest.h"
#include "absl/status/status.h"
#include "absl/time/time.h"
#include "google/fhir/testutil/proto_matchers.h"
#include "proto/google/fhir/proto/stu3/codes.pb.h"
#include "proto/google/fhir/proto/stu3/datatypes.pb.h"
#include "proto/google/fhir/proto/stu3/resources.pb.h"

namespace google {
namespace fhir {

namespace {

using ::google::fhir::stu3::proto::Encounter;
using ::google::fhir::stu3::proto::Patient;
using ::google::fhir::stu3::proto::Reference;
using ::google::fhir::testutil::EqualsProto;

TEST(GetResourceIdFromReferenceTest, PatientId) {
  Reference r;
  r.mutable_patient_id()->set_value("123");
  auto got = google::fhir::GetResourceIdFromReference<Patient>(r);
  FHIR_CHECK_OK(got.status());
  EXPECT_EQ(got.value(), "123");
}

TEST(GetResourceIdFromReferenceTest, PatientIdAsUri) {
  Reference r;
  r.mutable_uri()->set_value("Patient/123");
  auto got = google::fhir::GetResourceIdFromReference<Patient>(r);
  FHIR_CHECK_OK(got.status());
  EXPECT_EQ(got.value(), "123");
}

TEST(GetResourceIdFromReferenceTest, EncounterId) {
  Reference r;
  r.mutable_encounter_id()->set_value("456");
  auto got = google::fhir::GetResourceIdFromReference<Encounter>(r);
  FHIR_CHECK_OK(got.status());
  EXPECT_EQ(got.value(), "456");
}

TEST(GetResourceIdFromReferenceTest, EncounterIdAsUri) {
  Reference r;
  r.mutable_uri()->set_value("Encounter/456");
  auto got = google::fhir::GetResourceIdFromReference<Encounter>(r);
  FHIR_CHECK_OK(got.status());
  EXPECT_EQ(got.value(), "456");
}

TEST(GetResourceIdFromReferenceTest, ReferenceIsEmpty) {
  Reference r;
  auto got = google::fhir::GetResourceIdFromReference<Encounter>(r);
  EXPECT_EQ(got.status(), ::absl::NotFoundError("Reference is not populated."));
}

TEST(GetResourceIdFromReferenceTest, ReferenceDoesNotMatch) {
  Reference r;
  r.mutable_uri()->set_value("Patient/123");
  auto got = google::fhir::GetResourceIdFromReference<Encounter>(r);
  EXPECT_EQ(got.status(), ::absl::InvalidArgumentError(
                              "Reference `Patient/123` does not point to a "
                              "resource of type `Encounter`"));
}

TEST(GetReferenceProtoToResourceTest, UnprofiledProto) {
  Encounter resource;
  resource.mutable_id()->set_value("foo");

  Reference expected;
  expected.mutable_encounter_id()->set_value("foo");
  EXPECT_THAT(*google::fhir::GetReferenceProtoToResource<Reference>(resource),
              EqualsProto(expected));
}

TEST(ReferencesTest, RelativeReferenceSplit) {
  Reference r;
  r.mutable_uri()->set_value("Patient/123");
  ASSERT_TRUE(google::fhir::SplitIfRelativeReference(&r).ok());
  EXPECT_EQ(r.patient_id().value(), "123");
  EXPECT_TRUE(r.uri().value().empty());
}

TEST(ReferencesTest, UnrecognizedRelativeReferenceStaysAbsolute) {
  Reference r;
  r.mutable_uri()->set_value("Garbage/123");
  ASSERT_TRUE(google::fhir::SplitIfRelativeReference(&r).ok());
  EXPECT_EQ(r.uri().value(), "Garbage/123");
}

}  // namespace
}  // namespace fhir
}  // namespace google
